core/environment/manager.go

```diff
@@ -483,12 +483,43 @@ func (envs *Manager) TeardownEnvironment(environmentId uid.ID, force bool) error
 	if env.CurrentState() == "RUNNING" {
 		endTime, ok := env.workflow.GetUserVars().Get("run_end_time_ms")
 		if ok && endTime == "" {
-			runEndTime := strconv.FormatInt(time.Now().UnixMilli(), 10)
-			env.workflow.SetRuntimeVar("run_end_time_ms", runEndTime)
+			runEndTime := time.Now()
+			runEndTimeS := strconv.FormatInt(runEndTime.UnixMilli(), 10)
+			env.workflow.SetRuntimeVar("run_end_time_ms", runEndTimeS)
+
+			the.EventWriterWithTopic(topic.Run).WriteEventWithTimestamp(&evpb.Ev_RunEvent{
+				EnvironmentId:    environmentId.String(),
+				RunNumber:        env.GetCurrentRunNumber(),
+				State:            env.Sm.Current(),
+				Error:            "",
+				Transition:       "TEARDOWN",
+				TransitionStatus: evpb.OpStatus_STARTED,
+				Vars:             nil,
+			}, runEndTime)
 		} else {
 			log.WithField("partition", environmentId.String()).
 				Debug("O2 End time already set before DESTROY")
 		}
+
+		endCompletionTime, ok := env.workflow.GetUserVars().Get("run_end_completion_time_ms")
```

Comment:
Do I understand correctly that any state changes in the controlled tasks will be perfomed only by the line?
```
taskmanMessage := task.NewEnvironmentMessage(taskop.ReleaseTasks, environmentId, tasksToRelease, nil)
```
i'm wondering what would be better of the two, to set an EOEOR which is in fact before we leave RUNNING, or to set EOEOR after the tasks have been transitioned to DONE... The later seems more correct to me, but I am not sure I even understand correctly what happens here.

Reply:
OK, I see. Initially I read the code as the `DESTROY` hooks are called after having destroyed the tasks. It's only the message for releasing the tasks that is created before `DESTROY` hooks are called, but the message is only used after...